osu.Server.Spectator/Hubs/ScoreProcessedSubscriber.cs

```diff
@@ -0,0 +1,182 @@
+        timer.Start();
+
+        subscriber = redis.GetSubscriber();
+        subscriber.Subscribe("osu-channel:score:processed", (_, message) => onMessageReceived(message));
+
+        logger = Logger.GetLogger(nameof(ScoreProcessedSubscriber));
+    }
+
+    private void onMessageReceived(string message)
+    {
+        try
+        {
+            var scoreProcessed = JsonConvert.DeserializeObject<ScoreProcessed>(message);
+
+            if (scoreProcessed == null)
+                return;
+
+            if (subscriptions.TryRemove(scoreProcessed.ScoreId, out var subscription))
+            {
+                using (subscription)
+                    subscription.InvokeAsync().Wait();
+            }
+
+            DogStatsd.Increment($"{statsd_prefix}.messages.delivered");
+        }
+        catch (Exception ex)
+        {
+            logger.Add($"Failed to process message {message}", LogLevel.Error, ex);
+            DogStatsd.Increment($"{statsd_prefix}.messages.dropped");
+        }
+    }
+
+    public async Task RegisterForNotificationAsync(string receiverConnectionId, int userId, long scoreToken)
+    {
+        try
+        {
+            using var db = databaseFactory.GetInstance();
+
+            long? scoreId = await db.GetScoreIdFromToken(scoreToken);
+
+            if (scoreId == null)
+            {
+                DogStatsd.Increment($"{statsd_prefix}.subscriptions.dropped");
+                return;
+            }
+
+            var subscription = new ScoreProcessedSubscription(receiverConnectionId, userId, scoreId.Value, spectatorHubContext);
+
+            // because the score submission flow happens concurrently with the spectator play finished flow,
+            // it is theoretically possible for the score processing to complete before the spectator hub had a chance to register for notifications.
+            // to cover off this possibility, check the database directly once.
+            if (await db.IsScoreProcessedAsync(scoreId.Value))
+            {
+                using (subscription)
+                    await subscription.InvokeAsync();
+                DogStatsd.Increment($"{statsd_prefix}.messages.delivered-immediately");
+                return;
+            }
+
+            subscriptions.TryAdd(scoreId.Value, subscription);
+            DogStatsd.Gauge($"{statsd_prefix}.subscriptions.total", subscriptions.Count);
+        }
+        catch (Exception ex)
+        {
+            logger.Add($"Failed to register connection {receiverConnectionId} for info about score {userId}:{scoreToken}", LogLevel.Error, ex);
+            DogStatsd.Increment($"{statsd_prefix}.subscriptions.failed");
+        }
+    }
+
+    private void purgeTimedOutSubscriptions()
+    {
+        var scoreIds = subscriptions.Keys.ToArray();
+        int purgedCount = 0;
+
+        foreach (var scoreId in scoreIds)
+        {
+            if (subscriptions.TryGetValue(scoreId, out var subscription) && subscription.TimedOut)
```

Comment:
One thing to watch out for is that `CancellationTokenSource.Token` throws an exception if the CTS has been disposed. There could theoretically be a small moment where this `TryGetValue()` retrieves the value on this thread, but then the `using (subscription)` above runs and only after that does the `subscription.TimedOut` check happen.

Reply:
I was aware of this, but when I checked the implementation, then I didn't expect an `IsCancellationRequested` check to throw after disposal. Maybe bad form relying on implementation details.

Reply:
No you're correct, I didn't know there was such a discrepancy between `Token` and `IsCancellationRequested`